### PR TITLE
[Nano] add openvino quantization using pot

### PR DESCRIPTION
**Pull request by zhentaocc**
steps:

- torch model -> ONNX -> OpenVINO IR
- torch dataloader -> OpenVINO dataloader
- torchmetric -> OpenVINO metric
- POT execution

### CODE REVIEW

File: python/nano/test/openvino/pytorch/test_openvino_quantize.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from torchvision.models.mobilenetv3 import mobilenet_v3_small
+import torch
+from torch.utils.data.dataset import TensorDataset
+from torch.utils.data.dataloader import DataLoader
+import os
+
+
+class TestOpenVINO(TestCase):
+    def test_trainer_trace_openvino(self):
+        trainer = Trainer(max_epochs=1)
```
Comment (jason-dai) [inline python/nano/test/openvino/pytorch/test_openvino_quantize.py:30]: why do we need `max_epochs=1` if we only want to `quantize`?

File: python/nano/test/openvino/pytorch/test_openvino_quantize.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        x = torch.rand((10, 3, 256, 256))
+        y = torch.ones((10, ), dtype=torch.long)
+
+        ds = TensorDataset(x, y)
+        dataloader = DataLoader(ds, batch_size=1)
+
+        optimized_model = trainer.quantize(model, accelerator='openvino',
+                                           calib_dataloader=dataloader,
+                                           metric=F1(10),
+                                           backend='pot')
```
Comment (jason-dai) [inline python/nano/test/openvino/pytorch/test_openvino_quantize.py:42]: do we need to specify `backend='pot'` if `accelerator='openvino'`?

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py [hunk 3]
```diff
@@ -355,8 +355,29 @@ def quantize(self, model,  # remove the type requirement for type checking
                                                quantized_onnx_model)
                 else:
                     return return_pl_model
+        elif backend == 'pot' or accelerator == 'openvino':
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:358]: shall we just automatically infer the backend: using pot for openvino, and inc for others?

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py [hunk 4]
```diff
@@ -355,8 +355,29 @@ def quantize(self, model,  # remove the type requirement for type checking
...
+            model_type = type(model).__name__
+            if not model_type == 'PytorchOpenVINOModel':
+                if not input_sample:
+                    # input_sample can be a dataloader
+                    input_sample = calib_dataloader
+                model = Trainer.trace(model,
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:364]: need to check the type of model in `trace`

File: python/nano/src/bigdl/nano/deps/openvino/pytorch/pytorch_openvino_model.py [hunk 5]
```diff
@@ -76,6 +78,21 @@ def _load(path):
...
+            higher_better=True,
+            drop_type="relative",
+            maximal_drop=0.999,
+            max_iter_num=1,
+            n_requests=None,
+            sample_size=300):
+        # convert torch metric/dataloader to openvino format
+        if metric:
+            metric = PytorchOpenVINOMetric(metric=metric, higher_better=higher_better)
+        dataloader = PytorchOpenVINODataLoader(dataloader, collate_fn=self.tensors_to_numpy)
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/deps/openvino/pytorch/pytorch_openvino_model.py:93]: need to check the type of metric & dataloader somewhere?

File: python/nano/src/bigdl/nano/deps/openvino/core/openvino_model.py [hunk 6]
```diff
@@ -14,8 +14,13 @@
 # limitations under the License.
 #
 from pathlib import Path
+from tempfile import TemporaryDirectory
 from openvino.runtime import Core
 from openvino.runtime.passes import Manager
+from openvino.tools.pot.graph import load_model, save_model
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/deps/openvino/core/openvino_model.py:20]: Instead of unconditionally importing pot, make it an optional dependency (i.e., users only need to install pot if they use it).

Comment (zhentaocc) [reply python/nano/src/bigdl/nano/deps/openvino/core/openvino_model.py:20]: POT is installed with `openvino-dev`. So we don't have to check if pot is installed as long as openvino is installed.

File: python/nano/src/bigdl/nano/deps/openvino/pytorch/pytorch_openvino_model.py (hunk 5 above)
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/deps/openvino/pytorch/pytorch_openvino_model.py:93]: added checks in `PytorchOpenVINOMetric `and `PytorchOpenVINODataLoader`

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py [hunk 7]
```diff
@@ -220,7 +221,6 @@ def quantize(self, model,  # remove the type requirement for type checking
                  accuracy_criterion: dict = {'relative': 0.99, 'higher_is_better': True},
                  approach='static',
                  method='fx',
-                 backend='inc',
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:223]: also need to remove the python doc item below

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py [hunk 8]
```diff
@@ -360,6 +360,9 @@ def trace(model: nn.Module,
                                             accelerator='onnxruntime', otherwise will be ignored.
         :return: Model with different acceleration(OpenVINO/ONNX Runtime).
         """
+        assert isinstance(model, nn.Module) and not isinstance(model, AcceleratedLightningModule),\
+            "Expect a nn.Module instance that is not traced or quantized\
+                but got type {}".format(type(model))
         if accelerator == 'openvino':
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:366]: what if `accelerator` value is invalid?

File: python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py (hunk 7 above)
Comment (zhentaocc) [reply python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:223]: removed

**Comment by zhentaocc** [comment]
One more modification on `Trainer` will be included in another PR, making most of the default value as `None`. The reason is that some arguments are not used in openvino in some conditions and I think we need to check users are not trying to expect any changes by passing values to these arguments.  
Before 
```python
def quantize(self, model,  # remove the type requirement for type checking
             precision='int8',
             accelerator=None,
             calib_dataloader: DataLoader = None,
             val_dataloader: DataLoader = None,
             metric: Optional[Metric] = None,
             accuracy_criterion: dict = {'relative': 0.99, 'higher_is_better': True},
             approach='static',
             method='fx',
             conf: Optional[str] = None,
             tuning_strategy='bayesian',
             timeout=0,
             max_trials=1,
             input_sample=None
             ):
```
After
```python
def quantize(self, model,  # remove the type requirement for type checking
             precision='int8',
             accelerator=None,
             calib_dataloader: DataLoader = None,
             val_dataloader: DataLoader = None,
             metric: Optional[Metric] = None,
             accuracy_criterion: dict = None,
             approach='static',
             method=None,
             conf: Optional[str] = None,
             tuning_strategy=None,
             timeout=None,
             max_trials=None,
             input_sample=None
             ):
```

**Comment by TheaperDeng** [comment]
Since we are supporting this usage in openvino.
```python
# Case1: Trace and quantize
openvino_model = trainer.trace(model, accelerator='openvino', input_sample=x)
optimized_model = trainer.quantize(openvino_model, accelerator='openvino',
                                   calib_dataloader=dataloader)
```
We may support this for onnxruntime latter.

**Comment by zhentaocc** [comment]
> Since we are supporting this usage in openvino.
> 
> ```python
> # Case1: Trace and quantize
> openvino_model = trainer.trace(model, accelerator='openvino', input_sample=x)
> optimized_model = trainer.quantize(openvino_model, accelerator='openvino',
>                                    calib_dataloader=dataloader)
> ```
> 
> We may support this for onnxruntime latter.

I think onnxruntime already support this.

**Comment by zhentaocc** [comment]
http://10.112.231.51:18888/job/BigDL-NanoTests-PR-Validation/120/

**Comment by jason-dai** [comment]
LGTM

**Comment by zhentaocc** [comment]
http://10.112.231.51:18888/job/BigDL-NanoTests-PR-Validation/135/
